src/pubsub/ua_pubsub_keystorage.h

```diff
@@ -0,0 +1,396 @@
+                                          UA_ServerCallback callback,
+                                          UA_Duration timeToNextMs,
+                                          UA_UInt64 *callbackID);
+
+/**
+ * @brief It takes the current Key data, divide it into singing key, encrypting key and
+ * keyNonce according to security policy associated with PubSub Group and  set it in
+ * channel context of the assocaited PubSub Group. In case of pubSubGroupId is
+ * UA_NODEID_NULL, all the Reader/WriterGroup's channelcontext are updated with matching
+ * SecurityGroupId.
+ *
+ * @param server The server object
+ * @param pubSubGroupId the nodeId of the Reader/WirterGroup whose channel context to be
+ * updated
+ * @param securityGroupId The identifier for the SecurityGroup
+ * @return UA_StatusCode return status code
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_importKeyToChannelContext(UA_Server *server, const UA_NodeId pubSubGroupId,
+                                              const UA_String securityGroupId);
+
+/**
+ * @brief The callback function to update the current key from keystorage in the server
+ * and import the key into channel context of the associated PubSub Group
+ *
+ * @param server the server object
+ * @param keyStorage the pointer to the keystorage
+ */
+void
+UA_moveToNextkeyCallback(UA_Server *server, UA_PubSubKeyStorage *keyStorage);
+
+/**
+ * @brief It updates/adds the current and future keys into the existing KeyStorage.
+ * If the currentKeyID is known to existing keyStorage, then it is set as the currentKey
+ * and any future keys are appended to the existing list. If the currentKeyId is not know
+ * then, existing keyList is discarded and replaced with the new list.
+ *
+ * @param server the server object
+ * @param keyStorage pointer to the keystorage
+ * @param currentKey the currentKey data
+ * @param currentKeyID the identifier of the current Key
+ * @param futureKeySize the size of the future key list
+ * @param futureKeys the pointer to the future keys list
+ * @param msKeyLifeTime the updated time to move to next key
+ * @return UA_StatusCode the return status
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_update(UA_Server *server, UA_PubSubKeyStorage *keyStorage,
+                           const UA_ByteString *currentKey, UA_UInt32 currentKeyID,
+                           const size_t futureKeySize, UA_ByteString *futureKeys,
+                           UA_Duration msKeyLifeTime);
+
+/**
+ * @brief KeyStorage must be referenced by atleast one PubSubGroup. This method checks if
+ * KeyStorage is referenced by more then 1 PubSubGroup, then it decreases the
+ * referenceCount by one. If the referenceCount is exactly one then removes the KeyStorage
+ * from the server and PubSubGroup.
+ *
+ * @param server that holds the keystorage and PubSubGroups0
+ * @param groupId of the PubSubGroup
+ */
+void
+UA_PubSubKeyStorage_removeKeyStorage(UA_Server *server, UA_String securityGroupId);
```

Comment:
In nearly all cases where this is used, you first do a lookup, then check if `ks != NULL`, then call removeKeyStorage.
Just forward the ks pointer to `removeKeyStorage` instead of doing a duplicate lookup.